### PR TITLE
Add extensions as trainer argument

### LINKED ISSUE #3212: Add extensions as trainer argument
In general in Chainer, the parts of the trainer build on one another by calling earlier parts as arguments, i.e. ``iterator = SerialIterator(train, 100)``.

``chainer.trainer`` does currently doesn't allow this structure with regard to extentions, instead requiring them to be called afterwards as arguments to ``trainer.extend()``. To allow for a similar structure, a new argument ``extend`` could be added to the ``trainer`` _init_ call.

```python
from chainer.training import extensions

trainer = chainer.trainer(updater, (50, 'epoch'), out='result', extend=[
    extensions.Evaluator(tester, model, device=gpu_id),
    extensions.dump_graph('main/loss'),
    extensions.snapshot(trigger=(20, 'epoch')),
...])
```

### PR DESCRIPTION
**Pull request by neka-nat**
resolve #3212

### CODE REVIEW

File: chainer/testing/training.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 
 def get_trainer_with_mock_updater(
-        stop_trigger=(10, 'iteration'), iter_per_epoch=10):
+        stop_trigger=(10, 'iteration'), iter_per_epoch=10, extend=[]):
```
Comment (beam2d) [inline chainer/testing/training.py]: I think the argument name should basically be a noun. In this case, `extensions=` is better.

File: chainer/training/trainer.py [hunk 2]
```diff
@@ -130,7 +132,7 @@ class Trainer(object):
 
     """
 
-    def __init__(self, updater, stop_trigger=None, out='result'):
+    def __init__(self, updater, stop_trigger=None, out='result', extend=[]):
```
Comment (beam2d) [inline chainer/training/trainer.py]: ditto

File: tests/chainer_tests/testing_tests/test_training.py [hunk 3]
```diff
@@ -6,15 +6,21 @@
...
 
 
+def _dummy_extension(trainer):
+    pass
+
+
 @testing.parameterize(*testing.product({
     'stop_trigger': [(5, 'iteration'), (5, 'epoch')],
     'iter_per_epoch': [0.5, 1, 1.5, 5],
+    'extend': [[], [_dummy_extension]]
```
Comment (beam2d) [inline tests/chainer_tests/testing_tests/test_training.py]: Rename it to `extensions` like the above ones.

**Comment by neka-nat** [comment]
Thank you for the review. I fixed them.

**Comment by unnonouno** [comment]
jenkins test this please

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**